src/main/java/org/apache/sysds/runtime/matrix/data/LibCommonsMath.java

```diff
@@ -295,11 +281,283 @@ private static MatrixBlock computeMatrixInverse(Array2DRowRealMatrix in) {
+			throw new DMLRuntimeException("Lanczos algorithm can only be done on a symmetric matrix.");
+		}
+
+		int num_Threads = 1;
+
+		int m = in.getNumRows();
+		MatrixBlock v0 = new MatrixBlock(m, 1, 0.0);
+		MatrixBlock v1 = MatrixBlock.randOperations(m, 1, 1.0, 0.0, 1.0, "UNIFORM", 0xC0FFEE);
+
+		// normalize v1
+		double v1_sum = v1.sum();
+		RightScalarOperator op_div_scalar = new RightScalarOperator(Divide.getDivideFnObject(), v1_sum, num_Threads);
+		v1 = v1.scalarOperations(op_div_scalar, new MatrixBlock());
+		UnaryOperator op_sqrt = new UnaryOperator(Builtin.getBuiltinFnObject(Builtin.BuiltinCode.SQRT), num_Threads, true);
+		v1 = v1.unaryOperations(op_sqrt, new MatrixBlock());
+		if(v1.sumSq() != 1.0)
+			throw new DMLRuntimeException("Lanczos algorithm: v1 not correctly normalized");
+
+		MatrixBlock T = new MatrixBlock(m, m, 0.0);
+		MatrixBlock TV = new MatrixBlock(m, 1, 0.0);
+		MatrixBlock w1;
+
+		ReorgOperator op_t = new ReorgOperator(SwapIndex.getSwapIndexFnObject(), num_Threads);
+		TernaryOperator op_minus_mul = new TernaryOperator(MinusMultiply.getFnObject(), num_Threads);
+		AggregateBinaryOperator op_mul_agg = InstructionUtils.getMatMultOperator(num_Threads);
+
+		MatrixBlock beta = new MatrixBlock(1, 1, 0.0);
+		for(int i = 0; i < m; i++) {
+			if(i == 0)
+				TV.copy(v1);
+			else
+				TV = TV.append(v1, new MatrixBlock(), true);
+
+			w1 = in.aggregateBinaryOperations(in, v1, op_mul_agg);
+			MatrixBlock w1_t = w1.reorgOperations(op_t, new MatrixBlock(), 0, 0, m);
+			MatrixBlock alpha = w1_t.aggregateBinaryOperations(w1_t, v1, op_mul_agg);
+			if(i < m - 1) {
+				w1 = w1.ternaryOperations(op_minus_mul, v1, alpha, new MatrixBlock());
+				w1 = w1.ternaryOperations(op_minus_mul, v0, beta, new MatrixBlock());
+				beta.setValue(0, 0, Math.sqrt(w1.sumSq()));
+				v0.copy(v1);
+				op_div_scalar = (RightScalarOperator) op_div_scalar.setConstant(beta.getDouble(0, 0));
+				v1 = w1.scalarOperations(op_div_scalar, new MatrixBlock());
+
+				T.setValue(i + 1, i, beta.getValue(0, 0));
+				T.setValue(i, i + 1, beta.getValue(0, 0));
+			}
+			T.setValue(i, i, alpha.getValue(0, 0));
+		}
+
+		MatrixBlock[] e = multiReturnOperations(T, "eigen");
+		e[1] = TV.aggregateBinaryOperations(TV, e[1], op_mul_agg);
+		return e;
+	}
+
+	/**
+	 * Function to perform the QR decomposition.
+	 * Input must be a square matrix.
+	 *
+	 * @param in matrix object
+	 * @return array of matrix blocks [Q, R]
+	 */
+	private static MatrixBlock[] computeQR2(MatrixBlock in) {
+		if(in.getNumRows() != in.getNumColumns()) {
+			throw new DMLRuntimeException("QR2 Decomposition can only be done on a square matrix. "
+				+ "Input matrix is rectangular (rows=" + in.getNumRows() + ", cols=" + in.getNumColumns() + ")");
+		}
+
+		int num_Threads = 1;
+		int m = in.rlen;
+
+		MatrixBlock A_n = new MatrixBlock(m, m, 0.0);
+		A_n.copy(in);
+
+		MatrixBlock Q_n = new MatrixBlock(m, m, 0.0);
+		for(int i = 0; i < m; i++) {
+			Q_n.setValue(i, i, 1.0);
```

Comment:
you can allocate the matrix with a specific value in one of the other constructors.

Reply:
We are not sure how to create an identity matrix using the existing MatrixBlock constructors.
We could create a double[] and use that with MatrixBlock.init(), but this might not be the solution you have in mind.

Reply:
Oh, i read it wrongly, i quickly saw it as a matrix you filled with a specific value.
for now lets keep it as it is, unless you observe that there is some bottleneck related to Q_n in the first iteration of the algorithm.